kvserver: factor out log storage from Replica

pav-kv opened the pull request:
This PR decouples the log storage from `Replica` into the `replicaLogStorage` type. This allows further extracting it to a separate package/API and rigorously testing it in isolation.

The last commit adds a randomized stress test for the new type. It generates a typical raft log workload which includes log appends with or without overlapping, and truncations. The operations are applied to both `replicaLogStorage` and the reference implementation of `raft.LogStorage` (`raft.MemoryStorage`), and the test makes sure the observed state matches between the two.

This exercises the whole raft log storage stack, which includes Pebble, sideloaded storage (not quite fully yet), and the entry cache. It also exercises the use of the two mutexes, and helps finding concurrency issues. For example, while writing this test, I used the `StateLoader` outside the `raftMu` critical section, which led to a bug that the test caught.

Example output:

```
replica_raftlog_test.go:41: seed: 11946316121606401527
replica_raftlog_test.go:122: cache loads: 315548
replica_raftlog_test.go:123: cache hits: 92935
replica_raftlog_test.go:124: read bytes: 11367237
replica_raftlog_test.go:129: term=101, log=(4154, 4173], committed=4160
```

Part of #136109

pav-kv committed:
kvserver: factor out log storage from Replica

This commit decouples the log storage into the replicaLogStorage type,
which allows further extracting it to a separate package/API and
rigorously testing it in isolation.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145585)
<!-- Reviewable:end -->

pav-kv commented:
@tbg @arulajmani How dow the plan sound?

Also, please validate the idea: see how the fields are moved to `replica_raftlog.go` from `replica.go` (inside `Replica` definition). This allows instantiating `replicaLogStorage` without a `Replica` (just need a couple of mutexes and a few supporting things like metrics), and then we can test it in isolation.

pav-kv commented on pkg/kv/kvserver/replica_raftlog.go:
```diff
@@ -24,12 +32,98 @@ import (
+	mux mutexPair
+
+	// mu contains the fields protected by mux.mu. Equivalent to Replica.mu.
+	mu struct {
```
Can inline the mutexes so that the usage looks the same:

```golang
mu struct {
	*syncutil.RWMutex
	// ...
}

r.mu.field = ...
```

pav-kv commented on pkg/kv/kvserver/replica_raftlog.go:
```diff
@@ -24,12 +32,98 @@ import (
+		// Hence, they are fine for metrics etc., but not for deciding whether we
+		// should create another pending truncation. For the latter, we compute
+		// the post-pending-truncation size using pendingLogTruncations.
+		raftLogSize int64
```
Rename all the fields, now that they are in a small scope.

pav-kv commented on pkg/kv/kvserver/replica_raftlog.go:
```diff
@@ -24,12 +32,98 @@ import (
-// TODO(pav-kv): make it a proper type, and integrate with the logstore package.
-type replicaLogStorage Replica
+// TODO(pav-kv): integrate better with the logstore package.
+type replicaLogStorage struct {
```
Add a constructor.

pav-kv commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -138,6 +138,28 @@ func (r *replicaRaftStorage) InitialState() (raftpb.HardState, raftpb.ConfState,
 	return hs, cs, nil
 }
 
+// TODO(pav-kv): remove the adapter methods.
```
One way to do it: don't have a single `raft.Storage` interface. Use `raft.LogStorage` and `raft.StateStorage` separately.

pav-kv commented on pkg/kv/kvserver/replica_raftlog.go:
```diff
@@ -24,12 +32,98 @@ import (
+}
 
 // asLogStorage returns the raft.LogStorage implementation of this replica.
 func (r *Replica) asLogStorage() *replicaLogStorage {
```
Can remove this.

pav-kv commented on pkg/kv/kvserver/replica.go:
```diff
@@ -381,25 +383,12 @@ type Replica struct {
 		// on-disk storage for sideloaded SSTables. Always non-nil.
 		// TODO(pav-kv): remove, since this is duplicated in logStorage.
```
Do this.

pav-kv committed:
kvserver: squash last entry ID, refresh comments

Epic: none
Release note: none

pav-kv committed:
kvserver: squash truncated state, refresh comments

Epic: none
Release note: none

pav-kv committed:
kvserver: rename log size fields, refresh comments

Epic: none
Release note: none

pav-kv commented on pkg/kv/kvserver/replica_raftlog.go:
```diff
@@ -24,12 +32,97 @@ import (
-// TODO(pav-kv): make it a proper type, and integrate with the logstore package.
-type replicaLogStorage Replica
+// TODO(pav-kv): integrate better with the logstore package.
+type replicaLogStorage struct {
```
Yes, after #145439 and a couple more supporting changes.

pav-kv replied:
Done.

pav-kv committed:
kvserver: remove duplicate SideloadStorage

Epic: none
Release note: none

pav-kv committed:
kvserver: clarify engines referenced by log storage

Epic: none
Release note: none

pav-kv committed:
kvserver: randomized test for replicaLogStorage

The test generates a typical raft log workload which includes log
appends with or without overlapping, and truncations. Applies it to both
replicaLogStorage and its reference implementation (raft.MemoryStorage),
and makes sure that the observed state matches between the two.

This exercises the whole raft log storage stack, which includes Pebble,
sideloaded storage, and the entry cache.

Epic: none
Release note: none

pav-kv commented:
The randomized test found a semi-bug pretty quickly:


```
Append: 1/1 2/1 3/1 4/1 5/1 6/1 7/1 8/1 9/1
Truncate: 1/1
Append: 8/2
Read (4,8]: 5/1 6/1 7/1 8/1
Expected: 5/1 6/1 7/1 8/2
```

Suspected reason (haven't confirmed but makes sense): [this](https://github.com/cockroachdb/cockroach/blob/133a122bdadd95dbcab8cf4741e09bb5c4c2965a/pkg/kv/kvserver/logstore/logstore.go#L338) cache truncation comes only after the storage write. So a stale cached entry can be read. This is the same hazard that I captured in a TODO. We don't hit it because raft wouldn't be trying to read this index. What's strange is that I mimicked this behaviour in the test - shouldn't have hit this failure.

pav-kv commented:
Fixed the bug (test-only): forgot to put `StateLoader` access under the mutex, and so the test would occasionally write garbage to Pebble.

pav-kv committed:
kvserver: use log storage fields through type

Epic: none
Release note: none

pav-kv committed:
kvserver: attach byte monitoring to replicaLogStorage

Epic: none
Release note: none

pav-kv committed:
kvserver: mv byte monitoring to raft log file

Epic: none
Release note: none

pav-kv commented:
@tbg Ready for a pass.

pav-kv commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,302 @@
+	for i := 0; i < 5; i++ {
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "read"},
+			func(ctx context.Context) { rt.readLoop(ctx, fork) }))
```
These readers hold `Replica.mu`. Might as well add a `raftMu`-based reader, since we now mostly use the log under `raftMu`. It's less interesting to test here because it won't be concurrent with the `writeLoop`.

pav-kv commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -501,26 +501,29 @@ func (r *Replica) handleLeaseResult(
+	r.asLogStorage().stagePendingTruncationRaftMuLocked(pt)
+}
+
+func (r *replicaLogStorage) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
```
Move this to `replica_raftlog_writes.go`, as well as other truncation-related code. For followups.

pav-kv commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,302 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplicaLogStorage(t *testing.T) {
```
Document what this test does, and cross-link it from the big comment which explains the `RawNode` / mutex semantics. This test is basically the primer on how to use the type.

pav-kv replied:
Done.

pav-kv commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,302 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplicaLogStorage(t *testing.T) {
```
Are tests multi-threaded by default? Need to double-check here, since one of the aims is to test for concurrency issues.

pav-kv replied:
```
fmt.Println(runtime.NumCPU())
fmt.Println(runtime.GOMAXPROCS(0))

prints:
10
10
```

pav-kv commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,303 @@
+func (rt *replicaLogStorageTest) append(ctx context.Context, app raft.StorageAppend) {
+	// While a write is in flight, raft will never attempt to read the affected
+	// log indices. In reality, there can be multiple writes in flight, and raft
+	// tracks the lowest "protected" index. We don't do it here for simplicity,
```
This sentence might be confusing. This is just saying that we don't do multiple in-flight writes. But we still need to "protect" the in-flight write from reads, hence the `readable` var. This is the equivalent of `unstable` - all the writes are underneath it.

tbg commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,303 @@
+	rt.mu.RLock()
+	committed := rt.mu.committed
+	rt.mu.RUnlock()
+	app := raft.StorageAppend{Responses: []raftpb.Message{{}}} // force sync
```
And why do we need to force a sync? Is it just that our implementation preconditions the sync on there being responses? If we didn't have responses, it would deduce that the storage writes don't need durability (vs directly looking at the fact that there are entries?)

Bit more words in a comment could be helpful.

pav-kv replied:
Added a comment and assertion.

tbg commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,303 @@
+	rt.mu.RLock()
+	committed := rt.mu.committed
+	rt.mu.RUnlock()
+	comp := rt.ms.Compacted()
```
Should we also pull all the information from the other storage and assert that they're equal?

tbg commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,303 @@
+		ctx:     context.Background(),
+		cache:   raftentry.NewCache(1 << 5), // stress eviction
+		onSync:  noopSyncCallback{},
+		metrics: newStoreMetrics(metric.TestSampleInterval),
```
Looks like this is near the only field preventing `replicaLogStorage` to move to a subpackage? Exciting!

pav-kv replied:
Yeah, pretty much. There is also a bunch of direct access to `replicaLogStorage` fields, but hopefully there is enough structure there to wrap it into the type's API. Or the fields can be made public worst case.

tbg reviewed (approved):
Nice - this came together quite well, as always thank you for setting up the reviewer-friendly commit history, much appreciated.

No serious comments, I wouldn't be at all opposed if you just merged this and addressed all the tweaks in a follow-up.

pav-kv commented on pkg/kv/kvserver/raft_log_truncator.go:
```diff
@@ -151,7 +151,7 @@ type pendingTruncation struct {
 	//   possible that the truncated state of this replica is already >
 	//   expectedFirstIndex. We don't actually set isDeltaTrusted=false for this
-	//   case since we will change Replica.raftLogSizeTrusted to false after
+	//   case since we will change logStorage.sizeTrustedmu.sizeTrusted to false after
```
Fixed this and a few more in this commit.

pav-kv commented on pkg/kv/kvserver/replica_raftlog_test.go:
```diff
@@ -0,0 +1,303 @@
+		return
+	}
+	// Prepare the batch.
+	b := rt.ls.ls.Engine.NewWriteBatch()
```
`Unindexed` is [not recommended](https://github.com/cockroachdb/cockroach/blob/acd03343f136d89a9e629c538df2ae1ba2bbb2d6/pkg/storage/engine.go#L987-L988), and `NewWriteBatch` is strictly better here since we only write. Or do you mean we should mimic what the `Replica` code does?

pav-kv replied:
After #144615, we almost migrated off the unindexed batches. I think all the truncation code now uses `NewWriteBatch`.

pav-kv commented:
TFTR!

bors r=tbg
